lib/portlayer/exec/container.go

```diff
@@ -244,3 +250,77 @@ func (c *Container) Remove(ctx context.Context) error {
 
 	return nil
 }
+
+// return a list of container attributes
+func List(ctx context.Context, sess *session.Session, all *bool) ([]Container, error) {
+
+	// for now we'll go to the infrastructure
+	// future iteration will utilize cache & event stream
+	moVMs, err := infraContainers(ctx, sess)
+	if err != nil {
+		return nil, err
+	}
+	// convert to container
+	containers := convertInfraContainers(moVMs, all)
+	return containers, nil
+}
+
+// get the containerVMs from infrastructure
+func infraContainers(ctx context.Context, sess *session.Session) ([]mo.VirtualMachine, error) {
+	var rp mo.ResourcePool
+	var vms []mo.VirtualMachine
+
+	// popluate the vm property of the vch resource pool
+	if err := Config.ResourcePool.Properties(ctx, Config.ResourcePool.Reference(), []string{"vm"}, &rp); err != nil {
+		log.Errorf("List failed to get %s resource pool child vms: %s", sess.Pool.Name(), err)
+		return nil, err
+	}
+
+	// We now have morefs for the pools VMs, so
+	//define attributes do we need from the VMs
+	attrib := []string{"config", "runtime.powerState", "summary"}
+
+	// populate the vm properties
+	sess.Retrieve(ctx, rp.Vm, attrib, &vms)
+
+	return vms, nil
+}
+
+// convert the infra containers to a container object
+func convertInfraContainers(vms []mo.VirtualMachine, all *bool) []Container {
+	var containerVMs []Container
+
+	for i := range vms {
+		// TODO: wrap these and other params in filtering fuction
+		if vms[i].Config.Uuid == Config.PortlayerUUID {
+			// don't want it
+			log.Debug("Skipping VCH VM...")
+			continue
+		}
+		// poweredOn or all states
+		if !*all && vms[i].Runtime.PowerState == types.VirtualMachinePowerStatePoweredOff {
+			// don't want it
+			log.Debugf("Skipping poweredOff VM %s", vms[i].Config.Name)
+			continue
+		}
+
+		container := &Container{ExecConfig: &metadata.ExecutorConfig{}}
+		source := extraconfig.OptionValueSource(vms[i].Config.ExtraConfig)
+		extraconfig.Decode(source, container.ExecConfig)
+
+		// set state & friendly status
+		if vms[i].Runtime.PowerState == types.VirtualMachinePowerStatePoweredOn {
+			container.State = StateRunning
+			container.Status = "running"
+		} else {
+			container.State = StateRunning
```

Comment:
Asking to learn: is `StateRunning` still valid for a container with status `stopped`?